sound/soc/sof/topology.c

```diff
@@ -2160,6 +2160,7 @@ static int sof_link_hda_process(struct snd_sof_dev *sdev,
 	}
 
 	config->hda.link_dma_ch = slot;
+	dai->dai_config = kmemdup(config, size, GFP_KERNEL);
```

Comment:
yes, could you add a comment for it ?  I know you don't call sof_set_dai_config, but you need to add the comment "set config for the DAI with name and direction matching"

Reply:
Always check for memory alloc issues!

Reply:
There is no memory leak here now, previous may cause memory leak. The problem is like:
suppose we have a dai list:
{
dai_name="Analog", direction="Playback";
dai_name="Analog", direction="Capture";
dai_name="Digital MIC", direction="Capture";
dai_name="Digital SSP", direction="Playback";
...
}
In the previous code(with the remove part), will do the kmemdup if the name matches, e.g "Analog", when we only set for dai with name "Analog" and direction "Playback", it will kmemdup two times the dai config as only check the name,  "Analog Playback" and "Analog Capture" are all set to the dai config for "Analog Playback". 
The next time when set for dai with name "Analog" and direction "Capture", still kmemdup two times the capture dai config for both "Analog Playback" and "Analog Capture". 
From the tx message to firmware, we can see message for capture send two times, while no tx message for playback.
This is why the bug happens.

Reply:
I talked about memory alloc and sorry mate but this can fail. you need to test all memory allocation. No discussion allowed on this topic.
Memory leaks happen when you don't free (which is a different problem).

Also I am not clear about your change. We have 2 routines which do something similar:

````c
static int sof_link_hda_process(struct snd_sof_dev *sdev,
				struct snd_soc_dai_link *link,
				struct sof_ipc_dai_config *config,
				int slot,
				int direction)
{
	struct sof_ipc_reply reply;
	u32 size = sizeof(*config);
	struct snd_sof_dai *dai;
	int found = 0;
	int ret;

	/* for hda link, playback and capture are supported by different
	 * dai in FW. Here get the dai_index of each dai and send config
	 * to FW. In FW, each dai sets config by dai_index
	 */
	list_for_each_entry(dai, &sdev->dai_list, list) {
		if (!dai->name)
			continue;

		if (strcmp(link->name, dai->name) == 0 &&
		    dai->comp_dai.direction == direction) {
			config->dai_index = dai->comp_dai.dai_index;
			found = 1;
			break;
		}
	}
````

and 
````c
/* set config for all DAI's with name matching the link name */
static int sof_set_dai_config(struct snd_sof_dev *sdev, u32 size,
			      struct snd_soc_dai_link *link,
			      struct sof_ipc_dai_config *config)
{
	struct snd_sof_dai *dai;

	list_for_each_entry(dai, &sdev->dai_list, list) {
		if (!dai->name)
			continue;

		if (strcmp(link->name, dai->name) == 0) {
			dai->dai_config = kmemdup(config, size, GFP_KERNEL);
			if (!dai->dai_config)
				return -ENOMEM;
		}
	}

	return 0;
}
````
It seems to me we have two routines doing the same thing and that if the direction needs to be  used it could be factored in with a boolean flag. I don't like copy/paste code with differences, it's a sure way to write spaghetti code.

Reply:
@plbossart  I discussed with  @ranj063,  one issue in our restore process is: each dai link is restored by each dai component, but there are two dai components(playback and capture) in one dai link. This makes one dai_link is restored two times not one. I will refine this algorithm.

Reply:
but for HDA, this algorithm is correct

Reply:
Now only dai link with type SSP and DMIC call sof_set_dai_config and use dai name to find the dai.
After remove sof_set_dai_config in sof_link_hda_load, the HDA dai link find the dai with name and direction in sof_link_hda_process, so I think this is not duplicate routine.
The problem is: Is that only HDA will have playback/capture with the same dai name?
If yes, I think this is OK to keep this. If not, that means SSP and other still will have  playback/capture with the same dai name, then probably, we need to change  sof_set_dai_config to use name and direction to find the dai.
@RanderWang @ranj063 how do you think?

Reply:
yes,  playback/capture in SSP have the same dai name